File: ios/ImageFilters/CropperViewManager.swift
import Foundation
import UIKit
import React

@objc(CropperViewManager)
class CropperViewManager: RCTViewManager {
    
    override func view() -> UIView! {
        return CropperView()
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc func getCropRect(_ node: NSNumber) {
        DispatchQueue.main.async {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? CropperView {
                let rect = view.getCurrentCropRect()
                // Return via callback or promise if needed
            }
        }
    }
    
    @objc func resetCropRect(_ node: NSNumber) {
        DispatchQueue.main.async {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? CropperView {
                view.resetCropRect()
            }
        }
    }
    
    @objc func setCropRect(_ node: NSNumber, cropRect: NSDictionary) {
        DispatchQueue.main.async {
            if let view = self.bridge.uiManager.view(forReactTag: node) as? CropperView {
                view.setCropRect(cropRect)
            }
        }
    }
}

// MARK: - CropperView

class CropperView: UIView {
    
    private var imageView: UIImageView!
    private var scrollView: UIScrollView!
    private var cropOverlay: CropOverlayView!
    private let imageLoader = ImageLoader.shared
    
    private var currentImage: UIImage?
    private var imageSize: CGSize = .zero
    
    // Events
    @objc var onCropRectChange: RCTDirectEventBlock?
    @objc var onGestureEnd: RCTDirectEventBlock?
    
    // Props
    @objc var sourceUri: String? {
        didSet {
            if sourceUri != oldValue {
                loadImage()
            }
        }
    }
    
    @objc var initialCropRect: NSDictionary? {
        didSet {
            if let rect = initialCropRect {
                applyCropRect(rect)
            }
        }
    }
    
    @objc var aspectRatio: NSNumber? {
        didSet {
            cropOverlay?.aspectRatio = aspectRatio?.doubleValue
        }
    }
    
    @objc var minCropSize: NSDictionary? {
        didSet {
            if let size = minCropSize,
               let width = size["width"] as? CGFloat,
               let height = size["height"] as? CGFloat {
                cropOverlay?.minCropSize = CGSize(width: width, height: height)
            }
        }
    }
    
    @objc var showGrid: Bool = true {
        didSet {
            cropOverlay?.showGrid = showGrid
        }
    }
    
    @objc var gridColor: NSString? {
        didSet {
            if let colorString = gridColor as String? {
                cropOverlay?.gridColor = UIColor(hexString: colorString) ?? .white
            }
        }
    }
    
    @objc var overlayColor: NSString? {
        didSet {
            if let colorString = overlayColor as String? {
                cropOverlay?.overlayColor = UIColor(hexString: colorString) ?? UIColor.black.withAlphaComponent(0.5)
            }
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupView()
    }
    
    private func setupView() {
        backgroundColor = .black
        
        // Setup scroll view for pan and zoom
        scrollView = UIScrollView(frame: bounds)
        scrollView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        scrollView.delegate = self
        scrollView.minimumZoomScale = 1.0
        scrollView.maximumZoomScale = 3.0
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.showsVerticalScrollIndicator = false
        addSubview(scrollView)
        
        // Setup image view
        imageView = UIImageView(frame: bounds)
        imageView.contentMode = .scaleAspectFit
        imageView.isUserInteractionEnabled = true
        scrollView.addSubview(imageView)
        
        // Setup crop overlay
        cropOverlay = CropOverlayView(frame: bounds)
        cropOverlay.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        cropOverlay.delegate = self
        addSubview(cropOverlay)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        scrollView.frame = bounds
        updateImageLayout()
        cropOverlay.frame = bounds
    }
    
    private func loadImage() {
        guard let uri = sourceUri else { return }
        
        Task {
            do {
                let image = try await imageLoader.loadImage(from: uri)
                await MainActor.run {
                    self.currentImage = image
                    self.imageSize = image.size
                    self.imageView.image = image
                    self.updateImageLayout()
                    self.cropOverlay.resetToFullImage()
                }
            } catch {
                print("Failed to load image: \(error)")
            }
        }
    }
    
    private func updateImageLayout() {
        guard let image = currentImage else { return }
        
        let imageSize = image.size
        let viewSize = bounds.size
        
        let widthRatio = viewSize.width / imageSize.width
        let heightRatio = viewSize.height / imageSize.height
        let scale = min(widthRatio, heightRatio)
        
        let scaledWidth = imageSize.width * scale
        let scaledHeight = imageSize.height * scale
        
        imageView.frame = CGRect(
            x: (viewSize.width - scaledWidth) / 2,
            y: (viewSize.height - scaledHeight) / 2,
            width: scaledWidth,
            height: scaledHeight
        )
        
        scrollView.contentSize = imageView.frame.size
    }
    
    func getCurrentCropRect() -> [String: Any] {
        return cropOverlay.getCropRect()
    }
    
    func resetCropRect() {
        cropOverlay.resetToFullImage()
    }
    
    func setCropRect(_ rect: NSDictionary) {
        applyCropRect(rect)
    }
    
    private func applyCropRect(_ rect: NSDictionary) {
        guard let x = rect["x"] as? CGFloat,
              let y = rect["y"] as? CGFloat,
              let width = rect["width"] as? CGFloat,
              let height = rect["height"] as? CGFloat else {
            return
        }
        
        cropOverlay.setCropRect(CGRect(x: x, y: y, width: width, height: height))
    }
}

// MARK: - UIScrollViewDelegate

extension CropperView: UIScrollViewDelegate {
    func viewForZooming(in scrollView: UIScrollView) -> UIView? {
        return imageView
    }
    
    func scrollViewDidEndZooming(_ scrollView: UIScrollView, with view: UIView?, atScale scale: CGFloat) {
        notifyGestureEnd()
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        if !decelerate {
            notifyGestureEnd()
        }
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        notifyGestureEnd()
    }
}

// MARK: - CropOverlayDelegate

extension CropperView: CropOverlayDelegate {
    func cropOverlayDidChange(_ overlay: CropOverlayView, cropRect: CGRect) {
        onCropRectChange?([
            "cropRect": [
                "x": cropRect.origin.x,
                "y": cropRect.origin.y,
                "width": cropRect.size.width,
                "height": cropRect.size.height
            ]
        ])
    }
    
    func notifyGestureEnd() {
        let rect = cropOverlay.getCropRect()
        onGestureEnd?(["cropRect": rect])
    }
}

// MARK: - CropOverlayView

protocol CropOverlayDelegate: AnyObject {
    func cropOverlayDidChange(_ overlay: CropOverlayView, cropRect: CGRect)
}

class CropOverlayView: UIView {
    
    weak var delegate: CropOverlayDelegate?
    
    var aspectRatio: Double?
    var minCropSize: CGSize = CGSize(width: 50, height: 50)
    var showGrid: Bool = true {
        didSet {
            setNeedsDisplay()
        }
    }
    var gridColor: UIColor = .white {
        didSet {
            setNeedsDisplay()
        }
    }
    var overlayColor: UIColor = UIColor.black.withAlphaComponent(0.5) {
        didSet {
            setNeedsDisplay()
        }
    }
    
    private var cropRect: CGRect = .zero
    private var isDragging = false
    private var dragStartPoint: CGPoint = .zero
    private var dragStartRect: CGRect = .zero
    private var activeHandle: CropHandle = .none
    
    private enum CropHandle {
        case none
        case topLeft, topRight, bottomLeft, bottomRight
        case top, bottom, left, right
        case center
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setup()
    }
    
    private func setup() {
        backgroundColor = .clear
        isUserInteractionEnabled = true
        
        let panGesture = UIPanGestureRecognizer(target: self, action: #selector(handlePan(_:)))
        addGestureRecognizer(panGesture)
    }
    
    override func draw(_ rect: CGRect) {
        super.draw(rect)
        
        guard let context = UIGraphicsGetCurrentContext() else { return }
        
        // Draw overlay
        context.setFillColor(overlayColor.cgColor)
        context.fill(rect)
        
        // Clear crop area
        context.setBlendMode(.clear)
        context.fill(cropRect)
        context.setBlendMode(.normal)
        
        // Draw crop rect border
        context.setStrokeColor(UIColor.white.cgColor)
        context.setLineWidth(2.0)
        context.stroke(cropRect)
        
        // Draw grid
        if showGrid {
            context.setStrokeColor(gridColor.withAlphaComponent(0.5).cgColor)
            context.setLineWidth(1.0)
            
            // Vertical lines
            let thirdWidth = cropRect.width / 3
            for i in 1..<3 {
                let x = cropRect.minX + (thirdWidth * CGFloat(i))
                context.move(to: CGPoint(x: x, y: cropRect.minY))
                context.addLine(to: CGPoint(x: x, y: cropRect.maxY))
            }
            
            // Horizontal lines
            let thirdHeight = cropRect.height / 3
            for i in 1..<3 {
                let y = cropRect.minY + (thirdHeight * CGFloat(i))
                context.move(to: CGPoint(x: cropRect.minX, y: y))
                context.addLine(to: CGPoint(x: cropRect.maxX, y: y))
            }
            
            context.strokePath()
        }
        
        // Draw corner handles
        let handleSize: CGFloat = 30
        let handleOffset: CGFloat = 15
        context.setFillColor(UIColor.white.cgColor)
        
        let corners: [(CGPoint, CGFloat, CGFloat)] = [
            (CGPoint(x: cropRect.minX, y: cropRect.minY), 0, 0), // Top left
            (CGPoint(x: cropRect.maxX, y: cropRect.minY), -handleSize, 0), // Top right
            (CGPoint(x: cropRect.minX, y: cropRect.maxY), 0, -handleSize), // Bottom left
            (CGPoint(x: cropRect.maxX, y: cropRect.maxY), -handleSize, -handleSize), // Bottom right
        ]
        
        for (point, xOffset, yOffset) in corners {
            let handleRect = CGRect(
                x: point.x + xOffset,
                y: point.y + yOffset,
                width: handleSize,
                height: handleSize
            )
            context.fill(handleRect)
        }
    }
    
    @objc private func handlePan(_ gesture: UIPanGestureRecognizer) {
        let location = gesture.location(in: self)
        
        switch gesture.state {
        case .began:
            isDragging = true
            dragStartPoint = location
            dragStartRect = cropRect
            activeHandle = getHandle(at: location)
            
        case .changed:
            guard isDragging else { return }
            let translation = CGPoint(
                x: location.x - dragStartPoint.x,
                y: location.y - dragStartPoint.y
            )
            updateCropRect(with: translation)
            setNeedsDisplay()
            delegate?.cropOverlayDidChange(self, cropRect: cropRect)
            
        case .ended, .cancelled:
            isDragging = false
            activeHandle = .none
            
        default:
            break
        }
    }
    
    private func getHandle(at point: CGPoint) -> CropHandle {
        let handleSize: CGFloat = 44 // Touch target size
        
        // Check corners first
        if point.distance(to: CGPoint(x: cropRect.minX, y: cropRect.minY)) < handleSize {
            return .topLeft
        }
        if point.distance(to: CGPoint(x: cropRect.maxX, y: cropRect.minY)) < handleSize {
            return .topRight
        }
        if point.distance(to: CGPoint(x: cropRect.minX, y: cropRect.maxY)) < handleSize {
            return .bottomLeft
        }
        if point.distance(to: CGPoint(x: cropRect.maxX, y: cropRect.maxY)) < handleSize {
            return .bottomRight
        }
        
        // Check edges
        if abs(point.x - cropRect.minX) < handleSize && point.y > cropRect.minY && point.y < cropRect.maxY {
            return .left
        }
        if abs(point.x - cropRect.maxX) < handleSize && point.y > cropRect.minY && point.y < cropRect.maxY {
            return .right
        }
        if abs(point.y - cropRect.minY) < handleSize && point.x > cropRect.minX && point.x < cropRect.maxX {
            return .top
        }
        if abs(point.y - cropRect.maxY) < handleSize && point.x > cropRect.minX && point.x < cropRect.maxX {
            return .bottom
        }
        
        // Check center
        if cropRect.contains(point) {
            return .center
        }
        
        return .none
    }
    
    private func updateCropRect(with translation: CGPoint) {
        var newRect = dragStartRect
        
        switch activeHandle {
        case .topLeft:
            newRect.origin.x += translation.x
            newRect.origin.y += translation.y
            newRect.size.width -= translation.x
            newRect.size.height -= translation.y
            
        case .topRight:
            newRect.origin.y += translation.y
            newRect.size.width += translation.x
            newRect.size.height -= translation.y
            
        case .bottomLeft:
            newRect.origin.x += translation.x
            newRect.size.width -= translation.x
            newRect.size.height += translation.y
            
        case .bottomRight:
            newRect.size.width += translation.x
            newRect.size.height += translation.y
            
        case .left:
            newRect.origin.x += translation.x
            newRect.size.width -= translation.x
            
        case .right:
            newRect.size.width += translation.x
            
        case .top:
            newRect.origin.y += translation.y
            newRect.size.height -= translation.y
            
        case .bottom:
            newRect.size.height += translation.y
            
        case .center:
            newRect.origin.x += translation.x
            newRect.origin.y += translation.y
            
        case .none:
            return
        }
        
        // Apply aspect ratio constraint
        if let ratio = aspectRatio {
            newRect = applyAspectRatio(to: newRect, ratio: CGFloat(ratio), handle: activeHandle)
        }
        
        // Apply minimum size constraint
        newRect.size.width = max(newRect.size.width, minCropSize.width)
        newRect.size.height = max(newRect.size.height, minCropSize.height)
        
        // Keep within bounds
        newRect = newRect.intersection(bounds)
        
        cropRect = newRect
    }
    
    private func applyAspectRatio(to rect: CGRect, ratio: CGFloat, handle: CropHandle) -> CGRect {
        var newRect = rect
        
        switch handle {
        case .topLeft, .topRight, .bottomLeft, .bottomRight:
            // Adjust to maintain aspect ratio
            let currentRatio = rect.width / rect.height
            if currentRatio > ratio {
                newRect.size.width = rect.height * ratio
            } else {
                newRect.size.height = rect.width / ratio
            }
            
        case .left, .right:
            newRect.size.height = rect.width / ratio
            
        case .top, .bottom:
            newRect.size.width = rect.height * ratio
            
        default:
            break
        }
        
        return newRect
    }
    
    func resetToFullImage() {
        cropRect = bounds
        setNeedsDisplay()
    }
    
    func getCropRect() -> [String: Any] {
        return [
            "x": cropRect.origin.x,
            "y": cropRect.origin.y,
            "width": cropRect.size.width,
            "height": cropRect.size.height
        ]
    }
    
    func setCropRect(_ rect: CGRect) {
        cropRect = rect
        setNeedsDisplay()
    }
}

// MARK: - Helper Extensions

extension CGPoint {
    func distance(to point: CGPoint) -> CGFloat {
        return sqrt(pow(x - point.x, 2) + pow(y - point.y, 2))
    }
}

extension UIColor {
    convenience init?(hexString: String) {
        let hex = hexString.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int = UInt64()
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            return nil
        }
        self.init(red: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: CGFloat(a) / 255)
    }
}

File: android/src/main/java/com/imagefilters/CropperViewManager.kt
package com.imagefilters

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import android.view.View
import androidx.appcompat.widget.AppCompatImageView
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.common.MapBuilder
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.uimanager.events.RCTEventEmitter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class CropperViewManager(private val reactContext: ReactApplicationContext) :
    SimpleViewManager<CropperView>() {
    
    companion object {
        const val REACT_CLASS = "CropperView"
    }
    
    override fun getName(): String = REACT_CLASS
    
    override fun createViewInstance(reactContext: ThemedReactContext): CropperView {
        return CropperView(reactContext)
    }
    
    @ReactProp(name = "sourceUri")
    fun setSourceUri(view: CropperView, sourceUri: String?) {
        view.setSourceUri(sourceUri)
    }
    
    @ReactProp(name = "initialCropRect")
    fun setInitialCropRect(view: CropperView, cropRect: ReadableMap?) {
        view.setInitialCropRect(cropRect)
    }
    
    @ReactProp(name = "aspectRatio")
    fun setAspectRatio(view: CropperView, aspectRatio: Double) {
        view.setAspectRatio(if (aspectRatio > 0) aspectRatio.toFloat() else null)
    }
    
    @ReactProp(name = "minCropSize")
    fun setMinCropSize(view: CropperView, size: ReadableMap?) {
        view.setMinCropSize(size)
    }
    
    @ReactProp(name = "showGrid")
    fun setShowGrid(view: CropperView, showGrid: Boolean) {
        view.setShowGrid(showGrid)
    }
    
    @ReactProp(name = "gridColor")
    fun setGridColor(view: CropperView, color: String?) {
        view.setGridColor(color)
    }
    
    @ReactProp(name = "overlayColor")
    fun setOverlayColor(view: CropperView, color: String?) {
        view.setOverlayColor(color)
    }
    
    override fun getExportedCustomDirectEventTypeConstants(): Map<String, Any>? {
        return MapBuilder.of(
            "onCropRectChange", MapBuilder.of("registrationName", "onCropRectChange"),
            "onGestureEnd", MapBuilder.of("registrationName", "onGestureEnd")
        )
    }
    
    override fun getCommandsMap(): Map<String, Int>? {
        return MapBuilder.of(
            "getCropRect", 1,
            "resetCropRect", 2,
            "setCropRect", 3
        )
    }
    
    override fun receiveCommand(view: CropperView, commandId: String, args: com.facebook.react.bridge.ReadableArray?) {
        when (commandId) {
            "getCropRect" -> view.getCurrentCropRect()
            "resetCropRect" -> view.resetCropRect()
            "setCropRect" -> {
                args?.getMap(0)?.let { view.setCropRect(it) }
            }
        }
    }
}

@SuppressLint("ClickableViewAccessibility")
class CropperView(private val context: ThemedReactContext) : View(context) {
    
    private val imageLoader = ImageLoader(context)
    private val viewScope = CoroutineScope(Dispatchers.Main)
    
    private var currentBitmap: Bitmap? = null
    private var imageRect = RectF()
    private var cropRect = RectF()
    
    // Gesture detection
    private val scaleGestureDetector: ScaleGestureDetector
    private val gestureDetector: GestureDetector
    
    // Props
    private var aspectRatio: Float? = null
    private var minCropSize = PointF(50f, 50f)
    private var showGrid = true
    private var gridColor = Color.WHITE
    private var overlayColor = Color.argb(128, 0, 0, 0)
    
    // Drawing
    private val overlayPaint = Paint().apply {
        style = Paint.Style.FILL
    }
    private val strokePaint = Paint().apply {
        style = Paint.Style.STROKE
        strokeWidth = 4f
        color = Color.WHITE
    }
    private val gridPaint = Paint().apply {
        style = Paint.Style.STROKE
        strokeWidth = 2f
        color = Color.argb(128, 255, 255, 255)
    }
    private val handlePaint = Paint().apply {
        style = Paint.Style.FILL
        color = Color.WHITE
    }
    
    // Drag state
    private var dragHandle: CropHandle = CropHandle.NONE
    private var lastTouchX = 0f
    private var lastTouchY = 0f
    private var dragStartRect = RectF()
    
    private enum class CropHandle {
        NONE, TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT,
        TOP, BOTTOM, LEFT, RIGHT, CENTER
    }
    
    init {
        setBackgroundColor(Color.BLACK)
        setWillNotDraw(false)
        
        scaleGestureDetector = ScaleGestureDetector(context, ScaleListener())
        gestureDetector = GestureDetector(context, GestureListener())
        
        setOnTouchListener { _, event ->
            var handled = scaleGestureDetector.onTouchEvent(event)
            handled = gestureDetector.onTouchEvent(event) || handled
            
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    dragHandle = getHandleAtPoint(event.x, event.y)
                    lastTouchX = event.x
                    lastTouchY = event.y
                    dragStartRect.set(cropRect)
                    true
                }
                MotionEvent.ACTION_MOVE -> {
                    if (dragHandle != CropHandle.NONE) {
                        val dx = event.x - lastTouchX
                        val dy = event.y - lastTouchY
                        updateCropRect(dx, dy)
                        lastTouchX = event.x
                        lastTouchY = event.y
                        invalidate()
                        sendCropRectChangeEvent()
                        true
                    } else {
                        handled
                    }
                }
                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                    if (dragHandle != CropHandle.NONE) {
                        dragHandle = CropHandle.NONE
                        sendGestureEndEvent()
                    }
                    handled
                }
                else -> handled
            }
        }
    }
    
    fun setSourceUri(uri: String?) {
        if (uri.isNullOrEmpty()) return
        
        viewScope.launch {
            try {
                val bitmap = imageLoader.loadImage(uri)
                post {
                    // Recycle old bitmap on main thread to avoid race condition with drawing
                    val oldBitmap = currentBitmap
                    currentBitmap = bitmap
                    oldBitmap?.recycle()
                    
                    updateImageLayout()
                    resetCropRect()
                    invalidate()
                }
            } catch (e: Exception) {
                // Handle error
            }
        }
    }
    
    fun setInitialCropRect(rect: ReadableMap?) {
        rect?.let {
            val x = if (it.hasKey("x")) it.getDouble("x").toFloat() else 0f
            val y = if (it.hasKey("y")) it.getDouble("y").toFloat() else 0f
            val width = if (it.hasKey("width")) it.getDouble("width").toFloat() else 0f
            val height = if (it.hasKey("height")) it.getDouble("height").toFloat() else 0f
            
            cropRect.set(x, y, x + width, y + height)
            invalidate()
        }
    }
    
    fun setAspectRatio(ratio: Float?) {
        aspectRatio = ratio
    }
    
    fun setMinCropSize(size: ReadableMap?) {
        size?.let {
            val width = if (it.hasKey("width")) it.getDouble("width").toFloat() else 50f
            val height = if (it.hasKey("height")) it.getDouble("height").toFloat() else 50f
            minCropSize.set(width, height)
        }
    }
    
    fun setShowGrid(show: Boolean) {
        showGrid = show
        invalidate()
    }
    
    fun setGridColor(color: String?) {
        gridColor = parseColor(color) ?: Color.WHITE
        gridPaint.color = Color.argb(128, Color.red(gridColor), Color.green(gridColor), Color.blue(gridColor))
        invalidate()
    }
    
    fun setOverlayColor(color: String?) {
        overlayColor = parseColor(color) ?: Color.argb(128, 0, 0, 0)
        invalidate()
    }
    
    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        updateImageLayout()
        if (cropRect.isEmpty) {
            resetCropRect()
        }
    }
    
    private fun updateImageLayout() {
        val bitmap = currentBitmap
        if (bitmap == null || bitmap.isRecycled) {
            return
        }
        
        val viewWidth = width.toFloat()
        val viewHeight = height.toFloat()
        val imageWidth = bitmap.width.toFloat()
        val imageHeight = bitmap.height.toFloat()
        
        val scale = min(viewWidth / imageWidth, viewHeight / imageHeight)
        val scaledWidth = imageWidth * scale
        val scaledHeight = imageHeight * scale
        
        val left = (viewWidth - scaledWidth) / 2
        val top = (viewHeight - scaledHeight) / 2
        
        imageRect.set(left, top, left + scaledWidth, top + scaledHeight)
    }
    
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        
        val bitmap = currentBitmap
        // Check if bitmap exists and is not recycled
        if (bitmap == null || bitmap.isRecycled) {
            return
        }
        
        // Draw image
        canvas.drawBitmap(bitmap, null, imageRect, null)
        
        // Draw overlay (darken area outside crop)
        overlayPaint.color = overlayColor
        
        val path = Path().apply {
            addRect(0f, 0f, width.toFloat(), height.toFloat(), Path.Direction.CW)
            addRect(cropRect, Path.Direction.CCW)
        }
        canvas.drawPath(path, overlayPaint)
        
        // Draw crop rect border
        canvas.drawRect(cropRect, strokePaint)
        
        // Draw grid
        if (showGrid) {
            val thirdWidth = cropRect.width() / 3
            val thirdHeight = cropRect.height() / 3
            
            // Vertical lines
            for (i in 1..2) {
                val x = cropRect.left + (thirdWidth * i)
                canvas.drawLine(x, cropRect.top, x, cropRect.bottom, gridPaint)
            }
            
            // Horizontal lines
            for (i in 1..2) {
                val y = cropRect.top + (thirdHeight * i)
                canvas.drawLine(cropRect.left, y, cropRect.right, y, gridPaint)
            }
        }
        
        // Draw corner handles
        val handleSize = 40f
        drawHandle(canvas, cropRect.left, cropRect.top, handleSize)
        drawHandle(canvas, cropRect.right - handleSize, cropRect.top, handleSize)
        drawHandle(canvas, cropRect.left, cropRect.bottom - handleSize, handleSize)
        drawHandle(canvas, cropRect.right - handleSize, cropRect.bottom - handleSize, handleSize)
    }
    
    private fun drawHandle(canvas: Canvas, x: Float, y: Float, size: Float) {
        canvas.drawRect(x, y, x + size, y + size, handlePaint)
    }
    
    private fun getHandleAtPoint(x: Float, y: Float): CropHandle {
        val touchRadius = 60f
        
        // Check corners
        if (distance(x, y, cropRect.left, cropRect.top) < touchRadius) return CropHandle.TOP_LEFT
        if (distance(x, y, cropRect.right, cropRect.top) < touchRadius) return CropHandle.TOP_RIGHT
        if (distance(x, y, cropRect.left, cropRect.bottom) < touchRadius) return CropHandle.BOTTOM_LEFT
        if (distance(x, y, cropRect.right, cropRect.bottom) < touchRadius) return CropHandle.BOTTOM_RIGHT
        
        // Check edges
        if (abs(x - cropRect.left) < touchRadius && y > cropRect.top && y < cropRect.bottom) return CropHandle.LEFT
        if (abs(x - cropRect.right) < touchRadius && y > cropRect.top && y < cropRect.bottom) return CropHandle.RIGHT
        if (abs(y - cropRect.top) < touchRadius && x > cropRect.left && x < cropRect.right) return CropHandle.TOP
        if (abs(y - cropRect.bottom) < touchRadius && x > cropRect.left && x < cropRect.right) return CropHandle.BOTTOM
        
        // Check center
        if (cropRect.contains(x, y)) return CropHandle.CENTER
        
        return CropHandle.NONE
    }
    
    private fun distance(x1: Float, y1: Float, x2: Float, y2: Float): Float {
        return kotlin.math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2))
    }
    
    private fun updateCropRect(dx: Float, dy: Float) {
        val newRect = RectF(cropRect)
        
        when (dragHandle) {
            CropHandle.TOP_LEFT -> {
                newRect.left += dx
                newRect.top += dy
            }
            CropHandle.TOP_RIGHT -> {
                newRect.right += dx
                newRect.top += dy
            }
            CropHandle.BOTTOM_LEFT -> {
                newRect.left += dx
                newRect.bottom += dy
            }
            CropHandle.BOTTOM_RIGHT -> {
                newRect.right += dx
                newRect.bottom += dy
            }
            CropHandle.LEFT -> newRect.left += dx
            CropHandle.RIGHT -> newRect.right += dx
            CropHandle.TOP -> newRect.top += dy
            CropHandle.BOTTOM -> newRect.bottom += dy
            CropHandle.CENTER -> {
                newRect.offset(dx, dy)
            }
            CropHandle.NONE -> return
        }
        
        // Apply aspect ratio
        aspectRatio?.let { ratio ->
            newRect.set(applyAspectRatio(newRect, ratio, dragHandle))
        }
        
        // Apply minimum size
        val width = newRect.width()
        val height = newRect.height()
        if (width < minCropSize.x) {
            if (dragHandle in listOf(CropHandle.LEFT, CropHandle.TOP_LEFT, CropHandle.BOTTOM_LEFT)) {
                newRect.left = newRect.right - minCropSize.x
            } else {
                newRect.right = newRect.left + minCropSize.x
            }
        }
        if (height < minCropSize.y) {
            if (dragHandle in listOf(CropHandle.TOP, CropHandle.TOP_LEFT, CropHandle.TOP_RIGHT)) {
                newRect.top = newRect.bottom - minCropSize.y
            } else {
                newRect.bottom = newRect.top + minCropSize.y
            }
        }
        
        // Keep within image bounds
        if (newRect.left < imageRect.left) newRect.left = imageRect.left
        if (newRect.top < imageRect.top) newRect.top = imageRect.top
        if (newRect.right > imageRect.right) newRect.right = imageRect.right
        if (newRect.bottom > imageRect.bottom) newRect.bottom = imageRect.bottom
        
        cropRect.set(newRect)
    }
    
    private fun applyAspectRatio(rect: RectF, ratio: Float, handle: CropHandle): RectF {
        val newRect = RectF(rect)
        val currentRatio = rect.width() / rect.height()
        
        when (handle) {
            CropHandle.TOP_LEFT, CropHandle.TOP_RIGHT,
            CropHandle.BOTTOM_LEFT, CropHandle.BOTTOM_RIGHT -> {
                if (currentRatio > ratio) {
                    newRect.right = newRect.left + (newRect.height() * ratio)
                } else {
                    newRect.bottom = newRect.top + (newRect.width() / ratio)
                }
            }
            CropHandle.LEFT, CropHandle.RIGHT -> {
                newRect.bottom = newRect.top + (newRect.width() / ratio)
            }
            CropHandle.TOP, CropHandle.BOTTOM -> {
                newRect.right = newRect.left + (newRect.height() * ratio)
            }
            else -> {}
        }
        
        return newRect
    }
    
    fun getCurrentCropRect(): Map<String, Any> {
        return mapOf(
            "x" to cropRect.left,
            "y" to cropRect.top,
            "width" to cropRect.width(),
            "height" to cropRect.height()
        )
    }
    
    fun resetCropRect() {
        cropRect.set(imageRect)
        invalidate()
    }
    
    fun setCropRect(rect: ReadableMap) {
        setInitialCropRect(rect)
    }
    
    private fun sendCropRectChangeEvent() {
        val event = Arguments.createMap()
        val rectMap = Arguments.createMap()
        rectMap.putDouble("x", cropRect.left.toDouble())
        rectMap.putDouble("y", cropRect.top.toDouble())
        rectMap.putDouble("width", cropRect.width().toDouble())
        rectMap.putDouble("height", cropRect.height().toDouble())
        event.putMap("cropRect", rectMap)
        
        context.getJSModule(RCTEventEmitter::class.java)
            .receiveEvent(id, "onCropRectChange", event)
    }
    
    private fun sendGestureEndEvent() {
        val event = Arguments.createMap()
        val rectMap = Arguments.createMap()
        rectMap.putDouble("x", cropRect.left.toDouble())
        rectMap.putDouble("y", cropRect.top.toDouble())
        rectMap.putDouble("width", cropRect.width().toDouble())
        rectMap.putDouble("height", cropRect.height().toDouble())
        event.putMap("cropRect", rectMap)
        
        context.getJSModule(RCTEventEmitter::class.java)
            .receiveEvent(id, "onGestureEnd", event)
    }
    
    private fun parseColor(colorString: String?): Int? {
        if (colorString == null) return null
        return try {
            Color.parseColor(colorString)
        } catch (e: Exception) {
            null
        }
    }
    
    inner class ScaleListener : ScaleGestureDetector.SimpleOnScaleGestureListener() {
        override fun onScale(detector: ScaleGestureDetector): Boolean {
            // Can add zoom functionality here if needed
            return true
        }
    }
    
    inner class GestureListener : GestureDetector.SimpleOnGestureListener() {
        override fun onDown(e: MotionEvent): Boolean {
            return true
        }
    }
    
    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        currentBitmap?.let { bitmap ->
            if (!bitmap.isRecycled) {
                bitmap.recycle()
            }
        }
        currentBitmap = null
    }
}

